library/coqlib.ml

```diff
@@ -36,9 +201,64 @@ let find_reference locstr dir s =
       Pp.(str "cannot find " ++ Libnames.pr_path sp ++
           str "; maybe library " ++ DirPath.print dp ++
           str " has to be required first.")
+*)
 
 let coq_reference locstr dir s = find_reference locstr (coq::dir) s
 
+let table : (string, GlobRef.t Lazy.t) Hashtbl.t =
+  Hashtbl.create (2 * Array.length std_table)
+
+let init () =
+  Array.iter (fun (b, path, s) -> Hashtbl.add table b @@ lazy (find_reference "from_table" path s)) std_table
+
+(** Can throw Not_found *)
+let lib_ref ?(quiet=false) s =
+  try Lazy.force (Hashtbl.find table s)
+  with Not_found when not quiet ->
+    Feedback.msg_warning Pp.(str "not found in table: " ++ str s);
+    raise Not_found
+
+let add_ref s c =
+  Hashtbl.add table s (Lazy.from_val c)
+
+let cache_ref (_,(s,c)) =
+  add_ref s c
+
+let (inCoqlibRef : string * GlobRef.t -> Libobject.obj) =
+  let open Libobject in
+  declare_object { (default_object "COQLIBREF") with
+    cache_function = cache_ref;
+    load_function = (fun _ x -> cache_ref x);
+    classify_function = (fun o -> Keep o); (** FIXME figure out what to do when registering a functor argument field *)
```

Comment:
Is `Keep` the right behaviour? Who understands what Keep objects do?
BTW Keep is only used for generalizable variable currently.

Reply:
> Who understands what Keep objects do?

Not 100% sure, but I believe the difference is the behavior inside functors. `Keep` objects are attached to the functor itself, while `Subst` objects are attached to functors instances (and thus need to substitute names). Again, not really sure.

Reply:
Here is some behaviour for `Generalizable All Variables`:
~~~coq
Fail Definition foo `{x:A} := x.
(* control case 1 *)

Module Type E.   Global Generalizable All Variables. End E.

Fail Definition foo `{x:A} := x.
(* doesn't escape Module Type *)

Module F(A:E).
  Fail Definition foo `{x:A} := x.
  (* doesn't appear from the Module Type of an argument *)

  Global Generalizable All Variables.

  Fail Fail Definition foo `{x:A} := x.
  (* control case 2 *)
End F.

Fail Definition foo `{x:A} := x.
(* doesn't escape from a functor *)

Module G(A:E).
  Module M := F A.
  Import M.
  Fail Definition foo `{x:A} := x.
  (* doesn't escape from an applied functor *)
End G.

Module X. End X.

Include F. Include F X.
Fail Definition foo `{x:A} := x.
(* doesn't escape from an included functor (regardless of auto filling arguments) *)

Module M.
  Global Generalizable All Variables.
  Fail Fail Definition foo `{x:A} := x.
End M.

Fail Fail Definition foo `{x:A} := x.
(* does escape from regular Module (Generalizable takes effect at load_function) *)
~~~
If active in the toplevel module it also escapes through Require.

Are there any other module constructs?